src/nodes/dir.rs

```diff
@@ -518,6 +526,63 @@ impl Node for Dir {
         }))
     }
 
+    fn rename(&self, old_name: &OsStr, new_name: &OsStr) -> NodeResult<()> {
+        let mut state = self.state.lock().unwrap();
+
+        let old_path = Dir::get_writable_path(&mut state, old_name)?;
+        let new_path = Dir::get_writable_path(&mut state, new_name)?;
+
+        fs::rename(&old_path, &new_path)?;
+
+        let dirent = state.children.remove(old_name)
+            .expect("get_writable_path call above ensured the child exists");
+        dirent.node.set_underlying_path(&new_path);
+        state.children.insert(new_name.to_owned(), dirent);
+        Ok(())
+    }
+
+    fn rename_and_move_source(&self, old_name: &OsStr, new_dir: ArcNode, new_name: &OsStr)
+        -> NodeResult<()> {
+        debug_assert!(self.inode != new_dir.as_ref().inode(),
+            "Same-directory renames have to be done via `rename`");
+
+        let mut state = self.state.lock().unwrap();
+
+        let old_path = Dir::get_writable_path(&mut state, old_name)?;
+
+        let (old_name, dirent) = state.children.remove_entry(old_name)
+            .expect("get_writable_path call above ensured the child exists");
+        let result = new_dir.rename_and_move_target(&dirent, &old_path, new_name);
+        if result.is_err() {
+            // "Roll back" any changes we did to the current directory because the rename could not
+            // be completed on the target.
+            state.children.insert(old_name, dirent);
+        }
+        result
+    }
+
+    fn rename_and_move_target(&self, dirent: &Dirent, old_path: &Path, new_name: &OsStr)
+        -> NodeResult<()> {
+        // We are locking the target node while the source node is already locked, so there might be
```

Comment:
This will deadlock on any attempt to handle operations in parallel. You do need to mimic Go here, with a lock ordering constraint. Inode number is fine, as is any other totally ordered relation that is bijective with inodes.

Thankfully, visibility lets you reach into the other directory entry, and you should. Have a single `rename_and_move` function that takes the locks and then does the work.

Reply:
I believe `std::any::Any` will work here to allow you to downcast the Node.

Reply:
I had tried the downcast approach previously and couldn't get it to work. I did something like:

```
let new_dir = new_node.downcast::<Dir>();
```

but this complains about `downcast` not existing. I've seen somewhere that I may need to add an `as_any()` function to the `Node` to convert it to an `Any`, and then this could work here. Is that what has to happen?

(Eventually that's why I went with the current approach, to avoid the downcasts-that-didn't-work and having to expose `MutableDir` in the `Node` trait.)

By the way, I've added a new test (pushed here) to try to catch this potential deadlock... and haven't been able to. Even adding random sleeps after acquiring each lock is unable to catch the problem. I agree with you that the problem can exist though...

Reply:
The magic you are looking for is `trait Node : Any`.

Reply:
Is it possible there is some serialization happening at a higher level?

Reply:
Oh, duh... now that you mention it, the Rust FUSE library handles all requests sequentially... (Mind you, that's the primary reason it's way faster than the Go version.)

Reply:
Oh, my. That changes everything. You don't need `Arc` anywhere, many things can just be borrows, etc. (More particularly, your `ArcNode` can just be `&mut dyn Node`.

Reply:
Well, I think I still do need `Arc`s and stuff because of the reconfigurations, which happen from a separate thread.

I'll still try to get the `Any` solution working and the lock ordering fixed before merging this PR though.

Reply:
Fair enough. I tried the `Any` thingy and couldn't get it to work so going with this for now. Updated the TODO that was in the code to reflect reality and to explain that the deadlock would indeed happen.